### ISSUE TITLE
Capture of mu

### ISSUE THREAD

**Issue by gdcumming** [gfmoore/ESCI-JS#7]
±MoE around mu

This is good, but the two vertical lines should extend up to just under the latest sample dots—same height as the mu line—because we want to be able to see how the falling means lie relative to these two lines: for sigma known, red means are outside these lines, for sigma not known, approximately so.

Capture of mu by CIs

On all devices I’ve tried, this always seems to be on, whenever CIs are displayed, whether or not the mu line is displayed, whether or not the ‘capture of mu’ checkbox is ticked.

I suggest we should see capture numbers and % only when (1) CIs are displayed, and (2) the mu line is displayed, and (3) the ‘capture of mu’ checkbox is ticked. Click off 2 or 3 and the CIs should all revert to bright green.

Colours: I suggest both means and CIs should be: (1) bright green when capture not shown, or (2) darker green or red when capture is shown. (At present, CIs that capture are bright green, but their means are a very slightly darker green?)

Aha, you must always keep count of capture numbers and % separately for sigma known and unknown cases? Clever. So if click back and forth between sigma known and unknown we see the number and % captured change, if necessary. Nice! I’m guessing that this means that there’s no need to Clear before restarting with a different sigma known/unknown state selected. Correct?

Even more, if C% is changed, the number and % capturing seem to change to sensible values. Again, are you recording sufficient info for the whole run of samples to recalculate the values for number and % capturing, for the new value of C? If so, that fabulous! No need to Clear before restarting, even if C has been changed!

You mentioned that you have a buffer with all such info. Great. So ‘Clear’ must clear that buffer, as well as resetting the counts of samples to zero, I guess. Is that buffer of indefinitely large size? If not, we need some graceful way to cope if someone leaves the sampling on overnight and we get giant numbers of samples, for example if they wish to check how accurately the 95% (or other C value) predicts the number of CIs that do capture mu in a long run. It would certainly be nice to allow that, to check capture %, even if the advantages mentioned above of have a giant buffer can’t be supported beyond a certain number of samples in a run. (Like the ESCI limit that only the last 600 means are displayed in the mean heap.)

**Comment by gfmoore** [comment]
__±MoE around mu __

These are now same height as Mu line rather than growing with the heap.
I noticed that on Clear,  the ±MoE were cleared but the mu line wasn't - for consistency both left on after clear? Let me know.

__Capture of mu by CIs__

__Quote: "On all devices I’ve tried, this always seems to be on, whenever CIs are displayed, whether or not the mu line is displayed, whether or not the ‘capture of mu’ checkbox is ticked."__

I can't see this, did I fix it?  When I check it, then if MoE doesn't capture the mean it is red, otherwise green. If I uncheck it, always green.

 __Quote: "I suggest we should see capture numbers and % only when (1) CIs are displayed, and (2) the mu line is displayed, and (3) the ‘capture of mu’ checkbox is ticked. "__

Did you actually mean AND?  It made more sense (to me) to use OR but I can change this easily :)

__Quote: "Click off 2 (mu line) or 3 (capture of mu) and the CIs should all revert to bright green."__

Now the mu line doesn't have any effect (it shouldn't) on any colours, it's just a line.
When CI off there are no reds,  ever!!! (and it took me ages to get that code working properly!)
So hopefully this section ok now?

__Quote: "Colours: I suggest both means and CIs should be: (1) bright green when capture not shown, or (2) darker green or red when capture is shown. (At present, CIs that capture are bright green, but their means are a very slightly darker green?)"__

Originally I had the sigma know MoE bar colours a slightly lighter green so I could see that they were working when I was developing. Now all fills and bars are the same colour - a light green (actually "lawngreen"). When sample mean doesn't capture the mean then fill and bars coloured red.

Let me know if you want change, it shouldn't be quite easy to do this with the refactored code. (Though the code is getting  complex again...)


__Quote: "Aha, you must always keep count of capture numbers and % separately for sigma known and unknown cases? Clever. So if click back and forth between sigma known and unknown we see the number and % captured change, if necessary. Nice! I’m guessing that this means that there’s no need to Clear before restarting with a different sigma known/unknown state selected. Correct?"__

Indeed. I keep an array of all sample stats - just the M (xbar) and standard deviation of the sample. Everything else can be recalculated on the fly and JavaScript is so fast! (I also keep information attached to the svgs for the displayed sample means so that they can easily be switched as well. Of course when these drop of the end or hit the heap they disappear - hence the full list of all samples. It shouldn't fill up in a hurry!, but of course the bigger it is the longer things may take to recalculate. I'll have to recheck how many entry points there are and if these are just as a result of user interaction rather than any process.


__quote: " if someone leaves the sampling on overnight and we get giant numbers of samples, for example if they wish to check how accurately the 95% (or other C value) predicts the number of CIs that do capture mu in a long run..." __

Give it a go and see if it crashes!!!!  lol  (Turn the heap off!)

probably the field widths might be too small to display, but the percentage should be ok.

It does get very slow the longer it runs, not quite sure why yet. There's only that array that keeps getting added to, unless the heap is on and that could take huge amount of browser memory.


And on the mean heap... I think there's another issue for that, but if not I'll create one.

As I said it is probably easier to manage these issues if each one has it's own issue number, then I can focus on that one alone, even if they are related. I appreciate you going through and allocating them though. This is much easier than dropbox - which we don't seem to use now :)

Perhaps you could close this issue (if you can) and create a new issue for any changes. thanks.

**Comment by gdcumming** [comment]
Using 0.3.25

Mu line height: great
Boxes to click mu line, and (separately, independently) ±MoE both work well.

**Capture of mu by CIs**
This box still is ineffective. Whether ticked or not, I always get missing means and CIs red. As if it is always automatically turned on, and I can't turn it off. This happens for me on computer (Chrome), and iPhone and iPad (both Safari). 

Quote: "I suggest we should see capture numbers and % only when (1) CIs are displayed, and (2) the mu line is displayed, and (3) the ‘capture of mu’ checkbox is ticked."

I did mean AND. My thinking is that everything a student sees on screen at any moment should hang together, make sense. Deciding what accords with that aim is a judgment call, of course, So 'capture' makes most sense when (1) I have deliberately turned it on, and (2) I can see the mu line, because capture relates to that, and (3) CIs are displayed. (Having green and red means in the heap, of course without their CIs any longer displayed, needs to be an exception. Typically to be explained after understanding of capture and dropping mean is achieved, we hope.)

Therefore, if click off mu line box, or capture of mu box, or display CIs box, then we should no longer see capture counts and %.

**Colours**
Similarly, means and CIs should only appear red or darker green when all those 3 boxes are on (mu line, capture, CIs). Turn any one off and means (and intervals if CIs box still on) all become light green.

Quote: "Colours: I suggest both means and CIs should be: (1) bright green when capture not shown, or (2) darker green or red when capture is shown."

--yes please. 

In mean heap, I suggest that we also have colour of mean dots conform with the above. I.e. only show red-or-darkgreen when the conditions for capture display are met. Otherwise all means lightgreen. (This question didn't arise in ESCI because there I took the much more conservative position of not permitting CIs and mean heap to be displayed together. 'Clear' was forced when changing selection between those two. Having means with CIs drop onto the heap, as in -JS, now appeals to me, I confess.)

**Comment by gfmoore** [comment]
What does this 0.3.26 look like.
I finally understood what you meant about the capture of mu having no effect. You see it as being linked with CI and I saw it as independent. I think this is because they are in different panels.

I have to confess I don't quite understand your logic, but then you know exactly how you want to use it and have done for a long time :)
Personally I think if Capture of Mu is on, whether the u line is on or off should result in coloured heap and show capture statistics. The reason being is that you might not want the MoEs to be displayed and yet still see capturing take place. But hey, no problemo.

I hope I've got this right now, because trying to control all of this is too much for my tiny brain lol.

**Comment by gfmoore** [comment]
Oops, I've just realised you want the heap to colour match the dropping means. Okay.

**Comment by gfmoore** [comment]
Done, v 0.3.27

Do you want to change the colour of green? I'm not sure about the light green?

**Comment by gdcumming** [comment]
0.3.28

'morning Gordon,
You are moving so fast! Thanks for these new versions. Did I mention that I always *love* watching the green and red means piling so neatly into the mean heap. Also the CIs marching down the screen, also in their fine green and red.

Aha, I see. Sorry, I was not so clear about my assumed capture logic.

I think the Panel 6 then 7 logic is good. Think about introducing these ideas to students. After talking about means, dropping means (dance of the means), mean heap, we then talk about CIs on each mean, the dropping CIs (dance of the CIs). Only then do we mention the (definitional) idea of 95% of them capturing the (usually in real life, unknown) mu. Since we are in the computer, running a simulation, we can play God by knowing mu. So, with the CIs all visible, we turn on the mu line, then turn on capture, and--magic!--we see the means and CIs as either green or red. In real life, alas, CIs don't come conveniently coloured--we never can know mu--but one key lesson of all this playing in the computer is that, in real life, whenever we calculate or see a CI, IT MAY BE RED! Overall, roughly about 95% of the CIs we ever see will be red, but we'll never know which ones. Sad but true.

Then, given that experience playing with dances, we can go on and interpret the CI we calculate from our data in terms of the true value most likely being in our interval, and most likely somewhere the centre of our CI. Likelihood tails off smoothly both ways from the centre, out to and beyond the ends of the interval.

Back to your latest. Colours: I guess I'd prefer what I think is lawngreen (light and bright green) as the mean and CI colour when capture is turned off. All means and CIs are this colour. When capture is turned on, then lets have the red, as you use now, and maybe a green that is a little darker than lawngreen. To emphasise that these darker green means and CIs DO capture mu, whereas the red don't.

Main issue with the new logic is, I think, that I can't have CIs turned on, without both mu and capture being on. CIs on should give me bright green means and CIs. Turn on mu line and we see that. Turn on capture and all the means and CIs change colour. Turn off either or both mu line, and capture, and we revert to seeing CIs bright green.

(I started writing about #8 yesterday, was called away, lost what I'd written. I'll have another go at it: the complex logic of when to stop sampline, when to clear...)

**Comment by gfmoore** [comment]
_roughly about 95% of the CIs we ever see will be red,_ 
I think you meant green :)  Okay I think I get the pedagogy.

_I think, that I can't have CIs turned on, without both mu and capture being on_
I knew that having CI AND MuLine AND capture didn't make sense to me :)

Right So 
CI off   Muline off  capture mu off => just dropping lawn green means
CI on   Muline off  capture mu off => dropping lawn green means  and lawn green MoEs

CI off  Muline On Capture mu off  => get the mu line, dropping lawn green means  no MoEs
CI off  Muline Off Capture mu on => no effect just dropping lawn green means 
CI off Muline On Capture mu on => dropping means either dark green or red no MoEs

CI on  Muline on  Capture off => mu line on lawn green dropping means, no Moes
CI on Muline off Capture on => ???  lawn green means and moes
CI on Muline on Capture on =>mu line on, dropping means dark green or red MoEs dark green or red

Could you confirm this please before I make any changes.

**Comment by gdcumming** [comment]
0.3.30

roughly about 95% of the CIs we ever see will be red,
I think you meant green :) Okay I think I get the pedagogy.
>>>OK, you win, 5% will be red!

I think, that I can't have CIs turned on, without both mu and capture being on
I knew that having CI AND MuLine AND capture didn't make sense to me :)
>>>It's fine to have all those 3 on. In fact all three should be on if we are to show capture (red and darkgreen)

Right So
CI off Muline off capture mu off => just dropping lawn green means
CI on Muline off capture mu off => dropping lawn green means and lawn green MoEs
>>>yep (and means in heap are lawngreen in both cases)

CI off Muline On Capture mu off => get the mu line, dropping lawn green means no MoEs
CI off Muline Off Capture mu on => no effect just dropping lawn green means
>>>yep to both those (and lawngreen heap)

CI off Muline On Capture mu on => dropping means either dark green or red no MoEs
>>>no, I think CIs need to be visible if capture counts and %age, and red/darkgreen are to be visible. Unless all three are on, dropping means and means in heap should all be lawngreen

CI on Muline on Capture off => mu line on lawn green dropping means, no Moes
>>>yep

CI on Muline off Capture on => ??? lawn green means and moes
>>>yep

CI on Muline on Capture on =>mu line on, dropping means dark green or red MoEs dark green or red
>>>yep, also means in heap either red or darkgreen

>>>I'm thinking also that it may be good for 'Capture of mu' to clear automatically if CIs are turned off, or muline is turned off. If so, can we turn 'Capture of mu' on, if either or both of CIs and muline are currently off? It might be mysterious and annoying if it's not possible to click 'Capture of mu' on in that situation, even tho' clicking it on does not change the display--unless and until both those others are on. 

Could you confirm this please before I make any changes.

**Comment by gfmoore** [comment]
Okay I think I've covered all these now. I created 8 if statements to cope for all eventualities so if we decide to change then it is easy to do.
Personally, I did quite like to see the means turn green/red when capture of mu was checked on it's own. I quite liked to see the heap grow with green/red. Easier on the eye for testing that's for sure without the MoEs. No problem though :)

__I'm thinking also that it may be good for 'Capture of mu' to clear automatically if CIs are turned off, or muline is turned off__.

Okay, I've implemented that. (I hope I understood the requirement?)

btw how did you get the 3 vertical lines in your comments? Is that a markdown command?

Hopefully we can close this now? (unless I boo boo'd)
